Fix deprecated system colors to match same-as requirements in CSS Color 4

obyknovenius opened the pull request:
#### a396d027d05162da1014cc05f583b4688999beaf
<pre>
Fix deprecated system colors to match same-as requirements in CSS Color 4
<a href="https://bugs.webkit.org/show_bug.cgi?id=245609">https://bugs.webkit.org/show_bug.cgi?id=245609</a>

Reviewed by Aditya Keerthi.

CSS Color 4 deprecates some of the system colors
and adds a &quot;same-as&quot; requirement for each of them.

<a href="https://drafts.csswg.org/css-color-4/#deprecated-system-colors">https://drafts.csswg.org/css-color-4/#deprecated-system-colors</a>

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/system-color-compute-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/css/css-color/system-color-compute-expected.txt: Removed.
* LayoutTests/platform/glib/tables/mozilla/bugs/bug103533-expected.txt:
* LayoutTests/platform/gtk/TestExpectations:
* LayoutTests/platform/ios-wk2/TestExpectations:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/css/css-color/system-color-compute-expected.txt: Removed.
* LayoutTests/platform/ios/tables/mozilla/bugs/bug103533-expected.txt:
* LayoutTests/platform/mac/TestExpectations:
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/css/css-color/system-color-compute-expected.txt: Copied from LayoutTests/imported/w3c/web-platform-tests/css/css-color/system-color-compute-expected.txt.
* LayoutTests/platform/mac/tables/mozilla/bugs/bug103533-expected.txt:
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::systemColor const):
* Source/WebCore/rendering/RenderThemeAdwaita.cpp:
(WebCore::RenderThemeAdwaita::systemColor const):
* Source/WebCore/rendering/RenderThemeIOS.mm:
(WebCore::RenderThemeIOS::systemColor const):

Canonical link: <a href="https://commits.webkit.org/268124@main">https://commits.webkit.org/268124@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/079843ac1f8fcb222d9711596194e3bd827607b8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/18776 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/19117 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/19721 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/20642 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/17576 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/18969 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/22423 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/19255 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/19388 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/19001 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/19139 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/16352 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/21524 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/16361 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/17112 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/23557 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/17388 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/17284 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/21462 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/17883 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/15189 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/16946 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4457 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/21312 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/17721 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/rendering/RenderThemeIOS.mm:
```diff
@@ -1570,7 +1570,11 @@ static bool shouldUseConvexGradient(const Color& backgroundColor)
 
     auto& cache = colorCache(options);
-    return cache.systemStyleColors.ensure(cssValueID, [this, cssValueID, options] () -> Color {
+    auto it = cache.systemStyleColors.find(cssValueID);
+    if (it != cache.systemStyleColors.end())
```
Can you explain why this change is necessary? I'm not sure I see the difference between this and `ensure`.

obyknovenius replied:
This change prevents entering `cache.system.StyleColors.ensure()` twice.

This can happen because we call `RenderTheme::systemColor()` from `cache.systemStyleColors.ensure()`  when `globalCSSValueToSystemColorMap()` doesn't contain the color we're trying to resolve:

https://github.com/WebKit/WebKit/blob/b38121da42a0483fdb1bc13782f1c7a8059de100/Source/WebCore/rendering/RenderThemeIOS.mm#L1573-L1586

For deprecated colors (e.g. `CSSValueThreedface`) `RenderTheme::systemColor()` will call `RenderThemeIOS::systemColor()` again with "same-as" color (e.g. `CSSValueButtonface`).
https://github.com/WebKit/WebKit/blob/cb9b5c3f9215562642c1d48d92552c2586fdf474/Source/WebCore/rendering/RenderTheme.cpp#L1898-L1899

If this color is not yet cached we'll end up in `enusre()` again. This is "modifying-while-modifying" kind of problem. Which can lead to some really weird behavior.

pxlcoder replied:
Thanks!

pxlcoder commented on Source/WebCore/rendering/RenderThemeMac.mm:
```diff
@@ -473,7 +453,6 @@ static Color activeButtonTextColor()
         // The following colors would expose user appearance preferences to the web, and could be used for fingerprinting.
         // These are available only when the web view opts into the system appearance.
         case CSSValueWebkitFocusRingColor:
-        case CSSValueActiveborder:
```
I think all the changes to this file should be made behind a linked-on-or-after check, to avoid potentially breaking older apps.

This part of the change would need to be done by someone at Apple, as there are other dependencies. Happy to take that part on, but I would need to make that a separate PR.

(so this PR could leave failing `TestExpectations` for macOS only)

obyknovenius replied:
I have reverted changes in this file and added failing `TestExpectation` for macOS only.

pxlcoder commented on LayoutTests/platform/mac/TestExpectations:
```diff
@@ -2878,3 +2878,22 @@ webkit.org/b/261531 [ Monterey ] imported/w3c/web-platform-tests/css/css-transfo
+webkit.org/b/245347 imported/w3c/web-platform-tests/css/css-color/deprecated-sameas-021.html [ ImageOnlyFailure ]
+webkit.org/b/245347 imported/w3c/web-platform-tests/css/css-color/deprecated-sameas-022.html [ ImageOnlyFailure ]
+webkit.org/b/245347 imported/w3c/web-platform-tests/css/css-color/deprecated-sameas-023.html [ ImageOnlyFailure ]
+imported/w3c/web-platform-tests/css/css-color/system-color-compute.html
```
What's up with `imported/w3c/web-platform-tests/css/css-color/system-color-compute.html`? It's not in the original list, and doesn't have an expectation here.

obyknovenius replied:
This PR updates this test baseline with many `PASS` entries. But now it will fail on mac. And yes, I forgot to add `[ Failure ]`. It's fixed now.

pxlcoder replied:
In that case we should add a new expectation file at `LayoutTests/platform/mac/imported/w3c/web-platform-tests/css/css-color/system-color-compute-expected.txt` rather than mark it as a failure.

obyknovenius replied:
Done.

pxlcoder commented on Source/WebCore/rendering/RenderThemeIOS.mm:
```diff
@@ -1570,7 +1570,11 @@ static bool shouldUseConvexGradient(const Color& backgroundColor)
+    if (it != cache.systemStyleColors.end())
+        return it->value;
+
+    Color color = [this, cssValueID, options]() -> Color {
```
Nit: `auto color`

obyknovenius replied:
Fixed.

pxlcoder reviewed (approved):
Approved with [one comment](https://github.com/WebKit/WebKit/pull/16832#discussion_r1329134427) about a test.

obyknovenius committed:
Fix deprecated system colors to match same-as requirements in CSS Color 4
https://bugs.webkit.org/show_bug.cgi?id=245609

Reviewed by Aditya Keerthi.

CSS Color 4 deprecates some of the system colors
and adds a "same-as" requirement for each of them.

https://drafts.csswg.org/css-color-4/#deprecated-system-colors

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/system-color-compute-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/css/css-color/system-color-compute-expected.txt: Removed.
* LayoutTests/platform/glib/tables/mozilla/bugs/bug103533-expected.txt:
* LayoutTests/platform/gtk/TestExpectations:
* LayoutTests/platform/ios-wk2/TestExpectations:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/css/css-color/system-color-compute-expected.txt: Removed.
* LayoutTests/platform/ios/tables/mozilla/bugs/bug103533-expected.txt:
* LayoutTests/platform/mac/TestExpectations:
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/css/css-color/system-color-compute-expected.txt: Copied from LayoutTests/imported/w3c/web-platform-tests/css/css-color/system-color-compute-expected.txt.
* LayoutTests/platform/mac/tables/mozilla/bugs/bug103533-expected.txt:
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::systemColor const):
* Source/WebCore/rendering/RenderThemeAdwaita.cpp:
(WebCore::RenderThemeAdwaita::systemColor const):
* Source/WebCore/rendering/RenderThemeIOS.mm:
(WebCore::RenderThemeIOS::systemColor const):

Canonical link: https://commits.webkit.org/268124@main
